# colab setup


In [1]:
!pip install pip install git+https://github.com/ahmedbadr97/train-tracker

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/ahmedbadr97/train-tracker to /tmp/pip-req-build-q8bm269y
  Running command git clone -q https://github.com/ahmedbadr97/train-tracker /tmp/pip-req-build-q8bm269y
  Created wheel for train-tracker: filename=train_tracker-0.0.1-py3-none-any.whl size=7744 sha256=908a2a27ddcc00bb2d61a829d80749aee06d9f980142ac89a4d19d42e9a01a3f
  Stored in directory: /tmp/pip-ephem-wheel-cache-mfwci1y2/wheels/c0/62/5e/0351c11b3e5b24a5a6235943ff962a6a00e510e2391699df53
Successfully built train-tracker


In [2]:
!git clone https://github.com/ahmedbadr97/sentiment-analysis-RNN

Cloning into 'sentiment-analysis-RNN'...
remote: Enumerating objects: 171, done.
remote: Counting objects: 100% (171/171), done.
remote: Compressing objects: 100% (121/121), done.
remote: Total 171 (delta 84), reused 132 (delta 45), pack-reused 0
Receiving objects: 100% (171/171), 1.03 MiB | 11.77 MiB/s, done.
Resolving deltas: 100% (84/84), done.


In [ ]:
import config
#- config data --> git_username, email ,git_token, repository_name, kaggle_username, kaggle_key 
!git config --global user.name {config.git_username}
!git config --global user.email {config.email}


ModuleNotFoundError: ignored

In [ ]:
!git remote remove origin

!git remote add origin https://github.com/ahmedbadr97/sentiment-analysis-RNN

In [ ]:
!git pull origin main --force

remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 11 (delta 8), reused 8 (delta 5), pack-reused 0
Unpacking objects: 100% (11/11), done.
From https://github.com/ahmedbadr97/sentiment-analysis-RNN
 * branch            main       -> FETCH_HEAD
   7104443..c3935a6  main       -> origin/main
Updating 7104443..c3935a6
Fast-forward
 notebooks/Skip_Grams.ipynb | 10 +++++-----
 src/dataloader.py          | 27 +++++++++++++++++----------
 src/model_training.py      |  2 ++
 3 files changed, 24 insertions(+), 15 deletions(-)


In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
!unzip /content/drive/MyDrive/skipgram.zip -d /content/drive/MyDrive/projects/

unzip:  cannot find or open /content/drive/MyDrive/skipgram.zip, /content/drive/MyDrive/skipgram.zip.zip or /content/drive/MyDrive/skipgram.zip.ZIP.


In [ ]:
!zip -r /content/drive/MyDrive/skipgram.zip /content/drive/MyDrive/projects/sentiment-analysis-RNN

In [ ]:
!pip install gdown

In [ ]:
import gdown

In [ ]:
url="https://storage.googleapis.com/kaggle-data-sets/1875905/3063858/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20221202%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20221202T124244Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=05998512e2cb0a5988e3e21e206b01ef00e0b7bea305aa9a97cc711dfaba96aa68cfeb10628396ecb404474f2eb10563959884d971d333e9b49b4fc576dfffa2eecc091146b6d904e5f89b31414b0d6a0d65ee05447576de0a5f81a8165813f93a9467167d0d85f12b0a6b065ee03b3cf47f9f3a0b1624b5aad138ddfe5451b74aa09f25c35cef98e7e922c1340bb436f6a15070556d5bafb0130ba0fd3690672f6ee12a3375f4ff71a378a087c5f9418a561dcca8e384b521ab8d0c02d8c6ae431627dbbfa62880166f32c6f8d8c4c0cfc88f38e0835fb28e3e020bf308be765546455a71fa365443dd6ae26ab8039b3ffb95383b5793d52538cdc8a0929d91"
gdown.download(url,'/content/dataset.zip')

Downloading...
From: https://storage.googleapis.com/kaggle-data-sets/1875905/3063858/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20221202%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20221202T124244Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=05998512e2cb0a5988e3e21e206b01ef00e0b7bea305aa9a97cc711dfaba96aa68cfeb10628396ecb404474f2eb10563959884d971d333e9b49b4fc576dfffa2eecc091146b6d904e5f89b31414b0d6a0d65ee05447576de0a5f81a8165813f93a9467167d0d85f12b0a6b065ee03b3cf47f9f3a0b1624b5aad138ddfe5451b74aa09f25c35cef98e7e922c1340bb436f6a15070556d5bafb0130ba0fd3690672f6ee12a3375f4ff71a378a087c5f9418a561dcca8e384b521ab8d0c02d8c6ae431627dbbfa62880166f32c6f8d8c4c0cfc88f38e0835fb28e3e020bf308be765546455a71fa365443dd6ae26ab8039b3ffb95383b5793d52538cdc8a0929d91
To: /content/dataset.zip
100%|██████████| 21.6M/21.6M [00:00<00:00, 95.5MB/s]


'/content/dataset.zip'

In [ ]:
!unzip /content/dataset.zip -d {raw_dataset_dir}

Archive:  /content/dataset.zip
  inflating: /content/drive/MyDrive/projects/sentiment-analysis-RNN/dataset/raw/movie.csv  


In [4]:
%cd /content/sentiment-analysis-RNN

/content/sentiment-analysis-RNN


In [1]:
# Define Constants and imports
import os
from collections import Counter
import matplotlib.pyplot as plt
import pandas as pd
from string import punctuation
import src
from importlib import reload
import traintracker

reload(src)

raw_dataset_dir = "../dataset/raw"
prep_dataset_dir = "../dataset/preprocessed"

weights_dir="../model_weights/skipgram"
train_tracker_path="../train_tracker"

train_data_dir="../train_tracker"


# Introduction
In this notebook, I'll  implement the [Word2Vec algorithm](https://en.wikipedia.org/wiki/Word2vec) using the skip-gram architecture as a starter step for sentiment analysis creating first the words embeddings then use these embeddings for sentiment analysis

# Data Loading
- load and view the reviews dataset

In [4]:
os.listdir(raw_dataset_dir)

['labels.txt', 'reviews.txt', 'text8']

In [3]:
reviews_df=pd.read_csv(os.path.join(raw_dataset_dir,'movie.csv'))
reviews_df.info()

FileNotFoundError: [Errno 2] No such file or directory: '../dataset/raw\\movie.csv'

In [8]:
reviews_df.groupby(by='label').count()

text
label       
0      20019
1      19981

In [ ]:
reviews_df.head()

In [ ]:
print(reviews_df.iloc[0][0])
print(reviews_df.iloc[0][1])

In [11]:
# - change the reviews to one txt seprated by `\n` newline character
reviews_list=[]
labels_list=[]
size=len(reviews_df)
for i in range(size):
  reviews_list.append(reviews_df.iloc[i][0])
  
  labels_list.append("positive"  if reviews_df.iloc[i][1]==1 else "negative")
  if i<size-1:
    reviews_list.append('\n')
    labels_list.append('\n')


In [12]:
raw_reviews_txt="".join(reviews_list)
raw_labels_txt="".join(labels_list)
print(len(raw_reviews_txt.split('\n')))

40000


In [13]:
# saving labels to txt file
with open(os.path.join(prep_dataset_dir, "labels.txt"), 'w') as leables_file:
    leables_file.write(raw_labels_txt)

- reviews are line separated and the reviews also

## view words freqencies
- show word freqencies before any preprocessing

In [14]:
words_counter = Counter(raw_reviews_txt.split())


In [15]:
words_df = pd.DataFrame(words_counter.items(), columns=['word', 'count'])
words_df.sort_values(by=['count'], ascending=False, inplace=True)
print(f"no of unique words={len(words_df)}\nno of words ={words_df['count'].sum()}")

no of unique words=381542
no of words =9253570


In [16]:
# get 50 from top , mid and end
src.data_preprocessing.get_head_mid_tail(words_df,50)

word   count
8                  the  455003
51                   a  246023
6                  and  242041
53                  of  226965
27                  to  209659
...                ...     ...
198462  policewoman's,       1
198464         Palusky       1
198465           Kayle       1
198466         Timler,       1
381541         "Chains       1

[150 rows x 2 columns]

In [17]:
words_df.describe()

count
count  381542.000000
mean       24.253084
std      1240.617558
min         1.000000
25%         1.000000
50%         1.000000
75%         3.000000
max    455003.000000

- high standard deviation due the usage of words like 'the' and using punctuations in words make them counted as diffrent words
- 75% of the words count smaller than 3
- about 50% words appear once , may be misspelled words

In [18]:
words_df.median()

<ipython-input-18-0af9609fa379>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  words_df.median()


count    1.0
dtype: float64

- a words with count =1 appear from the midean to the end due to existance of punctations and the txt exist in thier lower and upper case 

# Data preprocessing
- remove punctuations , set all chars to lowercase
- remove noise words that appear once , that might be misspelling
- remove high frequent words that doesn't add thing to the neighboring word
- save the preprocessed data

**probability of word removal**
- Words that show up often such as "the", "of", and "for" don't provide much context to the nearby words. If we discard some of them, we can remove some of the noise from our data and in return get faster training and better representations. 
- This process is called subsampling by Mikolov.For each word wi) in the training set, 
- we'll discard it with probability given by 
- ![p_remove1.png](./assets/p_remove1.png)![p_remove2.png](./assets/p_remove2.png)
- p--> probability of discarding word
- t--> threshold of word count that the probability of discarding will begin to increase from `T/t`
- f--> frequency of a word
- T--> total number of words
- `set t=100 and T=500000=1e5 ` probability of removal will start to be grater than zero when the f=5000
- ![p_remove_curve.png]((./assets/p_remove_curve.png)

## remove invalid chars

In [41]:
prep_txt = src.remove_punctuations(raw_reviews_txt)


In [42]:
words_counter = Counter(prep_txt.split())

words_df = pd.DataFrame(words_counter.items(), columns=['word', 'count'])
words_df.sort_values(by=['count'], ascending=False, inplace=True)
print(f"no of unique words={len(words_df)}\nno of words ={words_df['count'].sum()}")

no of unique words=160310
no of words =9217166


no of unique words decreased from `381542`  to `160310` because we removed punctuation characters so word like `The. ` and `the?` will be `the` and will be counted as one word

In [43]:
def cnt_ranges(df,step,col_name):
  size=len(df)
  df=df.sort_values(by=col_name)
  col_idx=df.columns.get_loc(col_name)

  ranges_list=[]
  smaller_value=0
  bigger_value=step

  df_idx=0
  while(df_idx<size):
    cnt=0
    while(df_idx<size and df.iloc[df_idx,col_idx]<=bigger_value):
      cnt+=1
      df_idx+=1
    if cnt>0:
      ranges_list.append([f"{smaller_value} to {bigger_value}",cnt])
    smaller_value+=step
    bigger_value+=step
  return ranges_list


In [44]:
r=cnt_ranges(words_df,20,'count')

In [45]:
pd.DataFrame(r,columns=["range",'cnt'])

range     cnt
0             0 to 20  142824
1            20 to 40    6260
2            40 to 60    2771
3            60 to 80    1526
4           80 to 100    1059
..                ...     ...
397  213640 to 213660       1
398  230740 to 230760       1
399  256820 to 256840       1
400  256900 to 256920       1
401  531180 to 531200       1

[402 rows x 2 columns]

In [46]:
src.get_head_mid_tail(words_df,100)

word   count
8                        the  531191
6                        and  256916
47                         a  256838
49                        of  230759
23                        to  213654
...                      ...     ...
90777           pointyfinger       1
90776      handymancarpenter       1
90775              paragonbr       1
90774   signorelliscreenplay       1
160309           fishermenbr       1

[300 rows x 2 columns]

##  word importance by pos/neg apperance ratio

In [47]:
words_pos_neg_ratio = src.get_word_importance(prep_txt, labels=raw_labels_txt)
                                                                           
min_val,max_val=0,-1

for word,ratio in words_pos_neg_ratio.items():
    min_val=min(ratio,min_val)
    max_val=max(ratio,max_val)
print(f"min_val {min_val}")
print(f"max_val {max_val}")

min_val -4.127134385045092
max_val 4.727387818712341


In [48]:
raw_reviews_txt[:1000]

'I grew up (b. 1965) watching and loving the Thunderbirds. All my mates at school watched. We played "Thunderbirds" before school, during lunch and after school. We all wanted to be Virgil or Scott. No one wanted to be Alan. Counting down from 5 became an art form. I took my children to see the movie hoping they would get a glimpse of what I loved as a child. How bitterly disappointing. The only high point was the snappy theme tune. Not that it could compare with the original score of the Thunderbirds. Thankfully early Saturday mornings one television channel still plays reruns of the series Gerry Anderson and his wife created. Jonatha Frakes should hand in his directors chair, his version was completely hopeless. A waste of film. Utter rubbish. A CGI remake may be acceptable but replacing marionettes with Homo sapiens subsp. sapiens was a huge error of judgment.\nWhen I put this movie in my DVD player, and sat down with a coke and some chips, I had some expectations. I was hoping that

In [49]:
prep_txt[:1000]

'i grew up b 1965 watching and loving the thunderbirds all my mates at school watched we played thunderbirds before school during lunch and after school we all wanted to be virgil or scott no one wanted to be alan counting down from 5 became an art form i took my children to see the movie hoping they would get a glimpse of what i loved as a child how bitterly disappointing the only high point was the snappy theme tune not that it could compare with the original score of the thunderbirds thankfully early saturday mornings one television channel still plays reruns of the series gerry anderson and his wife created jonatha frakes should hand in his directors chair his version was completely hopeless a waste of film utter rubbish a cgi remake may be acceptable but replacing marionettes with homo sapiens subsp sapiens was a huge error of judgment \n when i put this movie in my dvd player and sat down with a coke and some chips i had some expectations i was hoping that this movie would contai

In [50]:
words_p_n_ratio_df=pd.DataFrame(words_pos_neg_ratio.items(),columns=['word','ratio']).sort_values(by=['ratio'])

In [51]:
words_p_n_ratio_df.describe()

ratio
count  161193.000000
mean        0.048999
std         0.818501
min        -4.127134
25%        -0.693147
50%         0.000000
75%         0.693147
max         4.727388

**Deciding the Ratio for noise**

In [52]:
print("common words")
print(f"pos and negative ratio for 'the' {words_pos_neg_ratio['those']}")
print(f"pos and negative ratio for 'him' {words_pos_neg_ratio['him']}")
print(f"pos and negative ratio for 'her' {words_pos_neg_ratio['her']}")
print(f"pos and negative ratio for 'is' {words_pos_neg_ratio['is']}")
print(f"pos and negative ratio for 'they' {words_pos_neg_ratio['they']}")
print(f"pos and negative ratio for 'did' {words_pos_neg_ratio['did']}")
print(f"pos and negative ratio for 'as' {words_pos_neg_ratio['as']}")
print(f"pos and negative ratio for 'a' {words_pos_neg_ratio['a']}")
print("\nPostive words")

print(f"pos and negative ratio for 'wonderful' {words_pos_neg_ratio['wonderful']}")
print(f"pos and negative ratio for 'amazing' {words_pos_neg_ratio['amazing']}")
print(f"pos and negative ratio for 'good' {words_pos_neg_ratio['good']}")
print(f"pos and negative ratio for 'excellent' {words_pos_neg_ratio['excellent']}")

print("\nnegative words")
print(f"pos and negative ratio for 'bad' {words_pos_neg_ratio['bad']}")
print(f"pos and negative ratio for 'worse' {words_pos_neg_ratio['worse']}")
print(f"pos and negative ratio for 'boring' {words_pos_neg_ratio['boring']}")

common words
pos and negative ratio for 'the' 0.15040138861478275
pos and negative ratio for 'him' 0.23241749635693773
pos and negative ratio for 'her' 0.24583923387574738
pos and negative ratio for 'is' 0.12302609115914889
pos and negative ratio for 'they' -0.2687558382621351
pos and negative ratio for 'did' -0.2160042141353108
pos and negative ratio for 'as' 0.23724039177675732
pos and negative ratio for 'a' 0.037063342877560294

Postive words
pos and negative ratio for 'wonderful' 1.5571694838883434
pos and negative ratio for 'amazing' 1.343095156017315
pos and negative ratio for 'good' 0.005656862045491339
pos and negative ratio for 'excellent' 1.4842294169777621

negative words
pos and negative ratio for 'bad' -1.362353199221966
pos and negative ratio for 'worse' -1.6799782168127269
pos and negative ratio for 'boring' -1.5131861129745143


In [53]:
ratio_threshold=0.27
is_common_word=(words_p_n_ratio_df['ratio']>-ratio_threshold)&(words_p_n_ratio_df['ratio']<ratio_threshold)
is_positive_word=(words_p_n_ratio_df['ratio']>ratio_threshold)
is_negative_word=(words_p_n_ratio_df['ratio']<-ratio_threshold)


common_words=words_p_n_ratio_df[is_common_word].sort_values(by='ratio')

positive_words=words_p_n_ratio_df[is_positive_word].sort_values(by='ratio')

negative_words=words_p_n_ratio_df[is_negative_word].sort_values(by='ratio')


In [54]:
print("common words ",common_words['word'].count())
print("positive words ",positive_words['word'].count())
print("negative words ",negative_words['word'].count())

common words  18124
positive words  75654
negative words  67415


In [55]:
n_words=300
common_words_range=src.get_head_mid_tail(common_words,n_words)
positive_words_range=src.get_head_mid_tail(positive_words,n_words)
negative_words_range=src.get_head_mid_tail(negative_words,n_words)

display(common_words_range)

word     ratio
3101      foot -0.269664
813     filthy -0.269664
45       would -0.269380
5529    bright -0.269255
2076      hide -0.269129
...        ...       ...
18671   downey  0.269664
19720     1986  0.269664
1528      told  0.269740
369    certain  0.269763
2939     entry  0.269824

[900 rows x 2 columns]

In [56]:
# add exceptional words
excep=['story','movie','good','film','movies']
for word in excep:
  words_pos_neg_ratio[word]=1

## remove word with hight freq and low freq

In [57]:
import src.data_preprocessing as prep
reload(prep)

<module 'src.data_preprocessing' from '/content/sentiment-analysis-RNN/src/data_preprocessing.py'>

In [62]:
filtered_prep_txt, noise_words,prob_drop_dist = prep.remove_noise(prep_txt, p_drop_dist_threshold=1e4, min_prob_drop=0.8, min_freq=20,min_rev_freq=5,common_words_threshold=0.27,word_importance_dict=words_pos_neg_ratio)

In [63]:
prob_drop_word_df=pd.DataFrame(prob_drop_dist.items(),columns=["word",'probabilty of drop']).sort_values(by=['probabilty of drop'],ascending=False)

In [64]:
prob_drop_word_df.head(150)

word  probabilty of drop
8             the            0.958344
6             and            0.940103
47              a            0.940094
49             of            0.936800
23             to            0.934318
...           ...                 ...
600          must            0.565623
1971        times            0.564955
470   interesting            0.563383
137        comedy            0.561931
390          done            0.559076

[150 rows x 2 columns]

In [35]:
noise_words

['replanted',
 '£5oo',
 'plotsso',
 'joydeep',
 'musseum',
 'gargon',
 'rebe',
 'phantasmagorical',
 'preferisco',
 'retaking',
 'talpiot',
 'seinfeldafter',
 'hosanna',
 'surmounts',
 'personalitymodern',
 'killerwhereas',
 'lightest',
 'kush',
 'trippers',
 'gametomovie',
 'chairwoman',
 'nopd',
 'toleration',
 'newagey',
 'eroticabr',
 'viciente',
 'expertonce',
 'unpharoh',
 'aoh',
 'gentrified',
 'twop',
 'selfreflexivity',
 'suspectdenniss',
 'rietty',
 'middleweight',
 'ce3k',
 'ambushers',
 'allthough',
 'talladega',
 'diary43',
 'joshuu',
 'chicagoset',
 'pising',
 'treasureshickey',
 'batlike',
 'dillonbr',
 'screenplaydirecting',
 'silkk',
 'peruses',
 'entrée',
 'soraia',
 'livefoes',
 'nonclassicbr',
 'appropriatebut',
 'notknowing',
 'dajani',
 'twogallons',
 'smolley',
 'carnbybr',
 'fuir',
 'astaire’s',
 'openedhe',
 'holstering',
 'capecloak',
 'factsthe',
 'muto',
 '11001001',
 'justoutofprison',
 'areweird',
 'brownsuch',
 'storycharactersatmosphere',
 'torturer',
 '

In [65]:
print(len(noise_words))
prep_txt=filtered_prep_txt

142367


In [66]:
words_counter = Counter(prep_txt.split())

words_df = pd.DataFrame(words_counter.items(), columns=['word', 'count'])
words_df.sort_values(by=['count'], ascending=False, inplace=True)
print(f"no of unique words={len(words_df)}\nno of words ={words_df['count'].sum()}")

no of unique words=17943
no of words =5189833


In [67]:
r=cnt_ranges(words_df,10,'count')

In [68]:
pd.DataFrame(r,columns=["range",'cnt']).head(100)

range   cnt
0       10 to 20   512
1       20 to 30  3929
2       30 to 40  2319
3       40 to 50  1606
4       50 to 60  1165
..           ...   ...
95    960 to 970    11
96    970 to 980     8
97    980 to 990     4
98   990 to 1000    17
99  1000 to 1010    11

[100 rows x 2 columns]

In [69]:
words_df[words_df['count']<100]

word  count
3901         christie     99
10866            1945     99
7894           akshay     99
6803            bombs     99
5024           motive     99
...               ...    ...
14749           nasal     20
17479          dealbr     20
16001  anthropologist     20
13613      songwriter     20
16294       sullivans     20

[12074 rows x 2 columns]

In [70]:
src.get_head_mid_tail(words_df,1000)

word  count
31              movie  66976
73               film  59411
62                his  45522
130              just  28030
129                if  26971
...               ...    ...
14749           nasal     20
17479          dealbr     20
16001  anthropologist     20
13613      songwriter     20
16294       sullivans     20

[3000 rows x 2 columns]

In [71]:
raw_reviews_txt[:250]

'I grew up (b. 1965) watching and loving the Thunderbirds. All my mates at school watched. We played "Thunderbirds" before school, during lunch and after school. We all wanted to be Virgil or Scott. No one wanted to be Alan. Counting down from 5 becam'

In [72]:
prep_txt[:250]

'grew up b 1965 watching loving thunderbirds my mates school watched we played thunderbirds before school during lunch after school we wanted virgil scott no wanted alan counting down 5 became art form took my children see movie hoping would get glimp'

In [73]:
# save preprocessed reviews
with open(os.path.join(prep_dataset_dir, "reviews.txt"), 'w') as reviews_file:
    reviews_file.write(prep_txt)

In [74]:
prep_words_counter = Counter(prep_txt.split())
prep_words_df = pd.DataFrame(prep_words_counter.items(), columns=['word', 'count'])
print(f"no of unique words={len(prep_words_df)}\nno of words ={prep_words_df['count'].sum()}")
prep_words_df.sort_values(by=['count'], ascending=False, inplace=True)

no of unique words=17943
no of words =5189833


In [75]:
prep_words_df.head(100)

word  count
31     movie  66976
73      film  59411
62       his  45522
130     just  28030
129       if  26971
..       ...    ...
65    should   7690
327  through   7646
120    those   7593
251       im   7563
783     back   7446

[100 rows x 2 columns]

- only `.` is removed with size `327192` , it was the second most frequent word , the std will decrease a little

In [76]:
prep_words_df.describe()

count
count  17943.000000
mean     289.239982
std     1373.175361
min       20.000000
25%       31.000000
50%       56.000000
75%      143.000000
max    66976.000000

In [77]:
prep_reviews_list = prep_txt.split('\n')
len(prep_reviews_list)

40000

In [78]:
prep_words_counter['séance']

0

# Skip-Gram Data loader
- Dataset Loader will load the txt data
- count the frequency for each word and total frequency which used in selecting the noise words which doesn't appear in the word context
- save a map for word to index and index to word , change txt to int
- iterate over the data , return each iteration the target which is the no of words around the given word with size `w` and the input will be the word repeated `w` times
    - ex if the txt is 'They celebrated his birthday with a big party.' and we are at the word 'birthday' and the window size = 4
        - input `[birthday,birthday,birthday,birthday]` target `[celebrated,his,with,a] `

- probability distribution of selecting a word


In [79]:
with open(os.path.join(prep_dataset_dir, 'reviews.txt')) as prep_file:
    prep_txt = prep_file.read()
prep_reviews_list = prep_txt.split('\n')
len(prep_reviews_list)

40000

In [80]:
rev_skip_gram_data = src.Word2VecDataset(prep_txt, window_size=5, no_noise_outputs=25, batch_size=256)

In [81]:
rev_skip_gram_data.save_word2int(prep_dataset_dir)

In [ ]:
for word in prep_words_counter.keys():
  if word not in rev_skip_gram_data.word2int:
    print(word)

In [82]:
word2intx=src.utils.load_json(os.path.join(prep_dataset_dir,"word2int.json"))


In [83]:
print(f"int value for amazing {word2intx['amazing']}")
print(f"int value for good {word2intx['good']}")

int value for amazing 444
int value for good 6


In [84]:
itrr = iter(rev_skip_gram_data)
word, target, noise = next(itrr)

In [85]:
print(f"words_in shape {word.shape}")
print(f"target_words shape {target.shape}")
print(f"noise_words shape {noise.shape}")

words_in shape torch.Size([1526])
target_words shape torch.Size([1526])
noise_words shape torch.Size([1526, 25])


In [86]:
rev_skip_gram_data.no_unique_words

17944

In [87]:
import torch
from torch import optim
import numpy as np
device = 'cuda' if torch.cuda.is_available() else 'cpu'

# instantiating the model
embedding_dim = 300
model = src.SkipGram(rev_skip_gram_data.no_unique_words, embedding_dim).to(device)

# Training

In [88]:

# using the loss that we defined
criterion = src.NegativeSamplingLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

epochs = 25

In [ ]:
latest_weights_path=traintracker.latest_weights_path(weights_dir)
print(latest_weights_path)
state_dict=torch.load(latest_weights_path)
model.load_state_dict(state_dict)

/content/drive/MyDrive/projects/sentiment-analysis-RNN/model_weights/skipgram/12_04 17_53 Train_(3.66341) .pt


<All keys matched successfully>

In [ ]:
notes="add nromalization from -3 to 3"

In [ ]:
src.skipgram_train(model=model,epochs=epochs,skip_gram_data=rev_skip_gram_data,device=device,optimizer=optimizer,criterion=criterion,train_data_dir=train_data_dir,weights_dir=weights_dir)

 epoch 1[..........]time remaining (m) = 9.44 Avg Train_Loss=30.11185708Epoch: 1/25
Loss:  12.038331985473633
film | dont, make, movie, movies, bad
think | evil, whole, actually, back, american
made | impressive, tell, no, well, same
them | if, make, we, remember, really
watch | film, even, –, see, making
really | left, where, women, girl, only
well | would, each, film, james, made
better | though, side, after, films, loved
former | underground, independent, passions, demons, seeks
younger | flaw, carey, jacket, effort, footage
college | somehow, fairly, spot, political, waybr
sweet | sounds, love, characters, does, argue
sadly | no, sculpture, statement, nearly, page
suddenly | quite, veteran, may, during, before
spend | best, bloody, eitherbr, minutes, dont
casting | lesson, soundtrack, completely, let, every
...

 epoch 1[=.........]time remaining (m) = 7.83 Avg Train_Loss=19.41773969Epoch: 1/25
Loss:  6.70511531829834
characters | through, always, films, god, well
scene | successfu

In [ ]:
src.train(model=model,epochs=epochs,skip_gram_data=rev_skip_gram_data,device=device,optimizer=optimizer,criterion=criterion,train_data_dir=train_data_dir,weights_dir=weights_dir)

 epoch 1[..........]time remaining (m) = 17.9 Avg Train_Loss=3.42570415Epoch: 1/15
Loss:  3.4633517265319824
your | want, yourself, please, then, our
best | worst, most, greatest, also, excellent
people | them, women, these, those, everyone
work | script, job, camera, works, director
something | anything, nothing, things, someone, least
still | made, though, years, while, also
actors | cast, actor, actresses, performances, acting
know | don, understand, think, do, say
century | th, twentieth, st, period, set
van | damme, dyke, sant, gus, helsing
accent | accents, english, southern, speaking, acting
tone | franchot, atmosphere, style, music, mood
starring | directed, plays, stars, starred, played
studio | mgm, studios, company, hollywood, disney
animated | animation, cartoon, disney, cartoons, bakshi
solid | excellent, fine, performances, great, supporting
...

 epoch 1[=.........]time remaining (m) = 15.73 Avg Train_Loss=3.43731387Epoch: 1/15
Loss:  3.4449784755706787
made | makes, mak

In [ ]:

saving_path="/content/drive/MyDrive"

full_path = '/content/drive/MyDrive/projects/sentiment-analysis-RNN/model_weights/skipgram/train_loss_4.21.pt'
torch.save(model.state_dict(), full_path)